filecoin-proofs/src/fr32_reader.rs

```diff
@@ -160,146 +108,44 @@ impl<R: io::Read> io::Read for Fr32Reader<R> {
-            read += current_read;
-
-            if current_read == 0 {
-                self.done = true;
-                break;
-            }
-        }
-
-        Ok(read)
-    }
-}
-
-fn set_bit(x: &mut u8, bit: usize) {
-    *x |= 1 << bit
-}
-
-use std::ops::{Deref, DerefMut};
-
-#[derive(Debug, Default, Clone, Copy)]
-struct Buffer {
-    data: u64,
-    /// Bits already consumed.
-    pos: u64,
-    /// Bits available.
-    avail: u64,
-}
+        let num_bytes = target.len();
 
-impl Deref for Buffer {
-    type Target = [u8; 8];
-
-    fn deref(&self) -> &Self::Target {
-        unsafe { &*(&self.data as *const u64 as *const [u8; 8]) }
-    }
-}
-
-impl DerefMut for Buffer {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        unsafe { &mut *(&mut self.data as *mut u64 as *mut [u8; 8]) }
-    }
-}
-
-impl Buffer {
-    /// How many bits are available to read.
-    #[inline]
-    pub fn available(&self) -> u64 {
-        self.avail - self.pos
-    }
-
-    pub fn reset_available(&mut self, bits: u64) {
-        self.pos = 0;
-        self.avail = bits;
-    }
-
-    /// Read a single bit at the current position.
-    pub fn read_bit(&mut self) -> bool {
-        let res = self.data & (1 << self.pos) != 0;
-        debug_assert!(self.available() >= 1);
-        self.pos += 1;
-        res
-    }
-
-    #[cfg(target_endian = "little")]
-    pub fn read_u8_range(&mut self, len: u64) -> u8 {
-        use bitintr::Bextr;
-        debug_assert!(self.available() >= len);
-        let res = self.data.bextr(self.pos, len) as u8;
-        self.pos += len;
-        res
-    }
-
-    #[cfg(target_endian = "little")]
-    pub fn read_u8(&mut self) -> u8 {
-        use bitintr::Bextr;
-        debug_assert!(self.available() >= 8);
-        let res = self.data.bextr(self.pos, 8) as u8;
-        self.pos += 8;
-        res
-    }
-
-    #[cfg(target_endian = "little")]
-    pub fn read_u16(&mut self) -> u16 {
-        debug_assert!(self.available() >= 16);
-
-        use bitintr::Bextr;
-        let res = self.data.bextr(self.pos, 16) as u16;
-        self.pos += 16;
-        res
-    }
+        let mut read = 0;
 
-    #[cfg(target_endian = "little")]
-    pub fn read_u16_into(&mut self, target: &mut [u8]) {
-        assert!(target.len() >= 2);
+        while read < num_bytes {
+            // load new block
+            if self.to_process == 0 {
+                let bytes_read = self.fill_buffer()?;
 
-        let value = self.read_u16().to_le_bytes();
-        target[0] = value[0];
-        target[1] = value[1];
-    }
+                // read all data, no new data in the buffer
+                if bytes_read == 0 {
+                    self.done = true;
+                    break;
+                }
 
-    #[cfg(target_endian = "little")]
-    pub fn read_u32(&mut self) -> u32 {
-        debug_assert!(self.available() >= 32);
+                self.process_block();
+                self.to_process += div_ceil(bytes_read * 8, 254);
```

Comment:
To process is in number of blocks. `bytes_read` is in `bytes`, so we multiply it by `8` to get bits. Because this is an input block its size is `254`, so we divide it by that, which is why I do not want to divide by `32`, even though this might be the same thing in the cases this code covers.